rbc/heavydb/geo_nested_array.py

```diff
@@ -0,0 +1,336 @@
+    "heavydb_geo_fromCoords_vec3",
+    "heavydb_geo_toCoords_vec",
+    "heavydb_geo_toCoords_vec2",
+    "heavydb_geo_toCoords_vec3",
+]
+
+import operator
+
+from llvmlite import ir
+from numba.core import cgutils, extending
+from numba.core import types as nb_types
+
+from rbc import typesystem
+from rbc.external import external
+
+from .allocator import allocate_varlen_buffer
+from .metatype import HeavyDBMetaType
+from .utils import as_voidptr, get_alloca
+
+i1 = ir.IntType(1)
+i8 = ir.IntType(8)
+i8p = i8.as_pointer()
+i8pp = i8p.as_pointer()
+i32 = ir.IntType(32)
+i64 = ir.IntType(64)
+i64p = i64.as_pointer()
+void = ir.VoidType()
+
+
+class GeoNestedArray(metaclass=HeavyDBMetaType):
+    """
+    Base class for GEO types.
+
+    .. note::
+        Geo columns should inherit from ``ColumnFlatBuffer`` in
+        ``column_flatbuffer.py``.
+
+    .. code-block:: c
+
+        {
+            int8_t* flatbuffer_;
+            int64_t index_[4];
+            int64_t n_;
+        }
+    """
+
+    def __len__(self) -> int:
+        """
+        """
+
+    def size(self) -> int:
+        """
+        """
+
+    def is_null(self) -> bool:
+        """
+        Return wether the Geo type is ``NULL``
+        """
+
+
+class GeoNestedArrayNumbaType(nb_types.Type):
+    def __init__(self, name):
+        super().__init__(name)
+        self.base_type = self.__typesystem_type__._params["base_type"]
+        self.item_type = self.__typesystem_type__._params["item_type"]
+
+
+class HeavyDBGeoNestedArray(typesystem.Type):
+    """Typesystem type class for HeavyDB buffer structures."""
+
+    @property
+    def numba_type(self):
+        return GeoNestedArrayNumbaType
+
+    @property
+    def type_name(self):
+        raise NotImplementedError()
+
+    @property
+    def item_type(self):
+        raise NotImplementedError()
+
+    def postprocess_type(self):
+        return self.params(shorttypename=self.type_name)
+
+    @property
+    def custom_params(self):
+        return {
+            "NumbaType": self.numba_type,
+            "name": self.type_name,
+            "base_type": self.type_name,
+            "item_type": self.item_type,
+        }
+
+    def tonumba(self, bool_is_int8=None):
+        flatbuffer_t = typesystem.Type.fromstring("int8_t* flatbuffer_")
+        index_t = typesystem.Type.fromstring("int64_t index_t")
+        n_t = typesystem.Type.fromstring("int64_t n_")
+        typ = typesystem.Type(
+            flatbuffer_t,
+            # int64_t index[4]
+            index_t,
+            index_t,
+            index_t,
+            index_t,
+            n_t,
+        )
+        typ.params(other=None, **self.custom_params)
+        return typ.tonumba(bool_is_int8=True)
+
+
+@extending.intrinsic
+def get_c_ptr(typingctx, sz, typ):
+    t = typ.dtype
+    sig = nb_types.CPointer(t)(sz, typ)
+
+    def codegen(context, builder, sig, args):
+        [sz, _] = args
+        # TODO: replace i64(8) by i64(sizeof(double))
```

Comment:
why `double`? IIUC, it should be related to the bitwidth of `typ`.

Reply:
You're right, it should be `i64(sizeof(typ_ptr))`, since we are creating a pointer here.